### PR TITLE
Improve area constraints processing

### LINKED ISSUE #879: Resizing tablet area using values breaks "Lock to usable area"
## Description
When resizing the tablet area with values, the area can go out-of-bounds.  
When the option is enabled, it is not possible when dragging the area with mouse.

https://user-images.githubusercontent.com/4985725/111660648-95424500-880e-11eb-8ec0-fa4848f42658.mp4

## System Information:

| Name             | Value |
| ---------------- | ----- |
| Operating System | Windows 10 |
| Software Version | v0.5.2.3 |
| Tablet           | Wacom CTL-472 |

### PR DESCRIPTION
**Pull request by X9VoiD**
- [x] Depends on #1287

## Changes

- Immediately apply "Lock to usable area" and "Enable aspect ratio lock" on boot when enabled.
- Immediately apply "Lock to usable area" whenever area size and/or position changes.
- Immediately apply "Lock to usable area" when the option is enabled.
- "Lock to usable area" now also locks maximum width and height. (Eto currently has a bug where the text box currently in focus won't receive the new constrained value, the area display shows the true value)

## Fixes

- Fixes #879

## Under-the-hood

- Move "Lock to usable area" logic to `AbsoluteModeEditor`.
- Hook and unhook dynamically to help reduce CPU usage during drag event.
- Use pre-allocated `EventArgs.Empty` instead where applicable, in `Area`-related code.

### CODE REVIEW

File: OpenTabletDriver.UX/Controls/Output/AbsoluteModeEditor.cs [hunk 1]
```diff
@@ -129,6 +198,71 @@ private void HandleAspectRatioLock(object sender, EventArgs e)
...
         }
 
+        private void HandleTabletAreaConstraint(object sender, EventArgs args)
+        {
+            ForceAreaConstraint((AreaDisplay)tabletAreaEditor, args);
+        }
+
+        private void HandleDisplayAreaConstraint(object sender, EventArgs args)
+        {
+            ForceAreaConstraint((AreaDisplay)displayAreaEditor, args);
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/Output/AbsoluteModeEditor.cs:208]: This cast isn't necessary, it gets casted later in the method anyway.

File: OpenTabletDriver.UX/Controls/Output/AbsoluteModeEditor.cs [hunk 2]
```diff
@@ -129,6 +198,71 @@ private void HandleAspectRatioLock(object sender, EventArgs e)
             }
         }
 
+        private void HandleTabletAreaConstraint(object sender, EventArgs args)
+        {
+            ForceAreaConstraint((AreaDisplay)tabletAreaEditor, args);
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/Output/AbsoluteModeEditor.cs:203]: Same here, unnecessary cast.

File: OpenTabletDriver.UX/Controls/Output/Area/AreaEditor.cs [hunk 3]
```diff
@@ -288,5 +288,10 @@ protected override void OnMouseDown(MouseEventArgs e)
                 }
             }
         }
+
+        public static explicit operator AreaDisplay(AreaEditor areaEditor)
+        {
+            return areaEditor.display;
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/Output/Area/AreaEditor.cs:295]: Not really sure how I feel about this, might just be better to expose the display field as a `get` only property.

Comment (X9VoiD) [reply OpenTabletDriver.UX/Controls/Output/Area/AreaEditor.cs]: Resolved.

**Review by InfinityGhost — APPROVED** [review]

**Pull request merged**